# Preparing LA Weather Dataset

We want to examine whether the weather has an effect on the type of crimes carried out in the time surrounding sports games. For example, if the weather is bad are more fans staying indoors to watch the games and in turn is domestic violence rising? Then alternatively to that, does good weather cause an increase in assault and vandalism, as more fans would be outdoor and attending the games?

In this notebook we will load in three datasets: one for temperature, one for weather description and one for wind speed. Each one contains data on 36 cities in the US and their columns is named after each one. We want to grade this weather so we can get an overall sense of how good or bad the weather was at certain times in the day. Each type of weather will contribute equally to determine what type of weather occurred. 

### Loading in the data

In [230]:
import os.path
import datetime
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline

The temperature dataset is loaded in. This contains hourly data of 36 different cities in the US. The datetime column is made the index column. 

In [231]:
if not os.path.exists( "../../data/raw/temperature.csv" ):
    print("Missing dataset file")

In [232]:
temp = pd.read_csv( "../../data/raw/temperature.csv", index_col="datetime", parse_dates=True)

In [233]:
temp.head()

Vancouver    Portland  San Francisco     Seattle  \
datetime                                                                 
2012-10-01 12:00:00         NaN         NaN            NaN         NaN   
2012-10-01 13:00:00  284.630000  282.080000     289.480000  281.800000   
2012-10-01 14:00:00  284.629041  282.083252     289.474993  281.797217   
2012-10-01 15:00:00  284.626998  282.091866     289.460618  281.789833   
2012-10-01 16:00:00  284.624955  282.100481     289.446243  281.782449   

                     Los Angeles   San Diego   Las Vegas     Phoenix  \
datetime                                                               
2012-10-01 12:00:00          NaN         NaN         NaN         NaN   
2012-10-01 13:00:00   291.870000  291.530000  293.410000  296.600000   
2012-10-01 14:00:00   291.868186  291.533501  293.403141  296.608509   
2012-10-01 15:00:00   291.862844  291.543355  293.392177  296.631487   
2012-10-01 16:00:00   291.857503  291.553209  293.381213  296.654466   

                     Albuquerque      Denver    ...      Philadelphia  \
datetime                                        ...                     
2012-10-01 12:00:00          NaN         NaN    ...               NaN   
2012-10-01 13:00:00   285.120000  284.610000    ...        285.630000   
2012-10-01 14:00:00   285.154558  284.607306    ...        285.663208   
2012-10-01 15:00:00   285.233952  284.599918    ...        285.756824   
2012-10-01 16:00:00   285.313345  284.592530    ...        285.850440   

                       New York    Montreal      Boston   Beersheba  \
datetime                                                              
2012-10-01 12:00:00         NaN         NaN         NaN         NaN   
2012-10-01 13:00:00  288.220000  285.830000  287.170000  307.590000   
2012-10-01 14:00:00  288.247676  285.834650  287.186092  307.590000   
2012-10-01 15:00:00  288.326940  285.847790  287.231672  307.391513   
2012-10-01 16:00:00  288.406203  285.860929  287.277251  307.145200   

                     Tel Aviv District       Eilat  Haifa  Nahariyya  \
datetime                                                               
2012-10-01 12:00:00                NaN  309.100000    NaN        NaN   
2012-10-01 13:00:00         305.470000  310.580000  304.4      304.4   
2012-10-01 14:00:00         304.310000  310.495769  304.4      304.4   
2012-10-01 15:00:00         304.281841  310.411538  304.4      304.4   
2012-10-01 16:00:00         304.238015  310.327308  304.4      304.4   

                     Jerusalem  
datetime                        
2012-10-01 12:00:00        NaN  
2012-10-01 13:00:00      303.5  
2012-10-01 14:00:00      303.5  
2012-10-01 15:00:00      303.5  
2012-10-01 16:00:00      303.5  

[5 rows x 36 columns]

Filtering the column from Los Angeles...

In [234]:
chi1 = temp.filter(items=['Los Angeles'])

In [235]:
chi1.head()

Los Angeles
datetime                        
2012-10-01 12:00:00          NaN
2012-10-01 13:00:00   291.870000
2012-10-01 14:00:00   291.868186
2012-10-01 15:00:00   291.862844
2012-10-01 16:00:00   291.857503

This column is renamed to 'Temperature'. In each dataset, whether it be the temperature, wind speed or description dataset, the column is named after the city. When we would join them, for this example we would be left with three columns named Los Angeles. 

In [236]:
chi1 = chi1.rename(columns={'Los Angeles': 'Temperature'})

In [237]:
chi1.head()

Temperature
datetime                        
2012-10-01 12:00:00          NaN
2012-10-01 13:00:00   291.870000
2012-10-01 14:00:00   291.868186
2012-10-01 15:00:00   291.862844
2012-10-01 16:00:00   291.857503

The wind speed dataset is now loaded in...

In [238]:
if not os.path.exists( "../../data/raw/wind_speed.csv" ):
    print("Missing dataset file")

The datetime column is again made the index..

In [239]:
temp1 = pd.read_csv( "../../data/raw/wind_speed.csv", index_col="datetime", parse_dates=True)

In [240]:
temp1.head()

Vancouver  Portland  San Francisco  Seattle  Los Angeles  \
datetime                                                                        
2012-10-01 12:00:00        NaN       NaN            NaN      NaN          NaN   
2012-10-01 13:00:00        0.0       0.0            2.0      0.0          0.0   
2012-10-01 14:00:00        0.0       0.0            2.0      0.0          0.0   
2012-10-01 15:00:00        0.0       0.0            2.0      0.0          0.0   
2012-10-01 16:00:00        0.0       0.0            2.0      0.0          0.0   

                     San Diego  Las Vegas  Phoenix  Albuquerque  Denver  \
datetime                                                                  
2012-10-01 12:00:00        NaN        NaN      NaN          NaN     NaN   
2012-10-01 13:00:00        0.0        0.0      2.0          4.0     4.0   
2012-10-01 14:00:00        0.0        0.0      2.0          4.0     4.0   
2012-10-01 15:00:00        0.0        0.0      2.0          4.0     3.0   
2012-10-01 16:00:00        0.0        0.0      2.0          4.0     3.0   

                       ...      Philadelphia  New York  Montreal  Boston  \
datetime               ...                                                 
2012-10-01 12:00:00    ...               NaN       NaN       NaN     NaN   
2012-10-01 13:00:00    ...               4.0       7.0       4.0     3.0   
2012-10-01 14:00:00    ...               4.0       7.0       4.0     3.0   
2012-10-01 15:00:00    ...               3.0       7.0       4.0     3.0   
2012-10-01 16:00:00    ...               3.0       7.0       4.0     3.0   

                     Beersheba  Tel Aviv District  Eilat  Haifa  Nahariyya  \
datetime                                                                     
2012-10-01 12:00:00        NaN                NaN    8.0    NaN        NaN   
2012-10-01 13:00:00        1.0                0.0    8.0    2.0        2.0   
2012-10-01 14:00:00        3.0                0.0    8.0    2.0        2.0   
2012-10-01 15:00:00        3.0                0.0    8.0    2.0        2.0   
2012-10-01 16:00:00        3.0                0.0    8.0    2.0        2.0   

                     Jerusalem  
datetime                        
2012-10-01 12:00:00        NaN  
2012-10-01 13:00:00        2.0  
2012-10-01 14:00:00        2.0  
2012-10-01 15:00:00        2.0  
2012-10-01 16:00:00        2.0  

[5 rows x 36 columns]

Again only interested in the Los Angeles column

In [241]:
chi2 = temp1.filter(items=['Los Angeles'])

In [242]:
chi2.head()

Los Angeles
datetime                        
2012-10-01 12:00:00          NaN
2012-10-01 13:00:00          0.0
2012-10-01 14:00:00          0.0
2012-10-01 15:00:00          0.0
2012-10-01 16:00:00          0.0

The Los Angeles column is renamed to 'Wind Speed'

In [243]:
chi2 = chi2.rename(columns={'Los Angeles': 'Wind Speed'})

In [244]:
chi2.head()

Wind Speed
datetime                       
2012-10-01 12:00:00         NaN
2012-10-01 13:00:00         0.0
2012-10-01 14:00:00         0.0
2012-10-01 15:00:00         0.0
2012-10-01 16:00:00         0.0

The weather description dataset is loaded in..

In [245]:
if not os.path.exists( "../../data/raw/weather_description.csv" ):
    print("Missing dataset file")

datetime is again made the index..

In [246]:
temp2 = pd.read_csv( "../../data/raw/weather_description.csv", index_col="datetime", parse_dates=True)

In [247]:
temp2.head()

Vancouver          Portland San Francisco  \
datetime                                                             
2012-10-01 12:00:00            NaN               NaN           NaN   
2012-10-01 13:00:00           mist  scattered clouds    light rain   
2012-10-01 14:00:00  broken clouds  scattered clouds  sky is clear   
2012-10-01 15:00:00  broken clouds  scattered clouds  sky is clear   
2012-10-01 16:00:00  broken clouds  scattered clouds  sky is clear   

                          Seattle   Los Angeles     San Diego     Las Vegas  \
datetime                                                                      
2012-10-01 12:00:00           NaN           NaN           NaN           NaN   
2012-10-01 13:00:00  sky is clear          mist  sky is clear  sky is clear   
2012-10-01 14:00:00  sky is clear  sky is clear  sky is clear  sky is clear   
2012-10-01 15:00:00  sky is clear  sky is clear  sky is clear  sky is clear   
2012-10-01 16:00:00  sky is clear  sky is clear  sky is clear  sky is clear   

                          Phoenix   Albuquerque         Denver  \
datetime                                                         
2012-10-01 12:00:00           NaN           NaN            NaN   
2012-10-01 13:00:00  sky is clear  sky is clear     light rain   
2012-10-01 14:00:00  sky is clear  sky is clear  broken clouds   
2012-10-01 15:00:00  sky is clear  sky is clear  broken clouds   
2012-10-01 16:00:00  sky is clear  sky is clear  broken clouds   

                          ...          Philadelphia    New York  \
datetime                  ...                                     
2012-10-01 12:00:00       ...                   NaN         NaN   
2012-10-01 13:00:00       ...         broken clouds  few clouds   
2012-10-01 14:00:00       ...         broken clouds  few clouds   
2012-10-01 15:00:00       ...         broken clouds  few clouds   
2012-10-01 16:00:00       ...         broken clouds  few clouds   

                            Montreal        Boston        Beersheba  \
datetime                                                              
2012-10-01 12:00:00              NaN           NaN              NaN   
2012-10-01 13:00:00  overcast clouds  sky is clear     sky is clear   
2012-10-01 14:00:00     sky is clear    few clouds     sky is clear   
2012-10-01 15:00:00     sky is clear    few clouds  overcast clouds   
2012-10-01 16:00:00     sky is clear    few clouds  overcast clouds   

                    Tel Aviv District          Eilat            Haifa  \
datetime                                                                
2012-10-01 12:00:00               NaN           haze              NaN   
2012-10-01 13:00:00      sky is clear           haze     sky is clear   
2012-10-01 14:00:00      sky is clear  broken clouds  overcast clouds   
2012-10-01 15:00:00      sky is clear  broken clouds  overcast clouds   
2012-10-01 16:00:00      sky is clear  broken clouds  overcast clouds   

                           Nahariyya        Jerusalem  
datetime                                               
2012-10-01 12:00:00              NaN              NaN  
2012-10-01 13:00:00     sky is clear     sky is clear  
2012-10-01 14:00:00     sky is clear  overcast clouds  
2012-10-01 15:00:00  overcast clouds  overcast clouds  
2012-10-01 16:00:00  overcast clouds  overcast clouds  

[5 rows x 36 columns]

Again, only looking at the Los Angeles column

In [248]:
chi3 = temp2.filter(items=['Los Angeles'])

In [249]:
chi3.head()

Los Angeles
datetime                         
2012-10-01 12:00:00           NaN
2012-10-01 13:00:00          mist
2012-10-01 14:00:00  sky is clear
2012-10-01 15:00:00  sky is clear
2012-10-01 16:00:00  sky is clear

This column is renamed to 'Description'

In [250]:
chi3 = chi3.rename(columns={'Los Angeles': 'Description'})

In [251]:
chi3.head()

Description
datetime                         
2012-10-01 12:00:00           NaN
2012-10-01 13:00:00          mist
2012-10-01 14:00:00  sky is clear
2012-10-01 15:00:00  sky is clear
2012-10-01 16:00:00  sky is clear

The Temperature, Wind Speed and Description columns are now concatted into the one dataframe. 

In [252]:
frames = [chi1, chi2, chi3]
df = pd.concat(frames, axis = 1)

In [253]:
df.head()

Temperature  Wind Speed   Description
datetime                                                  
2012-10-01 12:00:00          NaN         NaN           NaN
2012-10-01 13:00:00   291.870000         0.0          mist
2012-10-01 14:00:00   291.868186         0.0  sky is clear
2012-10-01 15:00:00   291.862844         0.0  sky is clear
2012-10-01 16:00:00   291.857503         0.0  sky is clear

### Dealing with Null values

We look at the null values in the new dataframe. 

In [254]:
df.isnull().sum()

Temperature    3
Wind Speed     1
Description    1
dtype: int64

Looking at the rows where Description is null... The only example is in the first row which is black in all fields so this row is deleted. 

In [255]:
print(df[df["Description"].isnull()])

                     Temperature  Wind Speed Description
datetime                                                
2012-10-01 12:00:00          NaN         NaN         NaN


In [256]:
df = df.dropna(subset = ['Description'])

Looking at the temperature fields that are null. 

In [257]:
print(df[df["Temperature"].isnull()])

                     Temperature  Wind Speed   Description
datetime                                                  
2013-03-11 07:00:00          NaN         0.0  sky is clear
2013-03-11 08:00:00          NaN         1.0  sky is clear


Looking at the day that consists of the null values. 

In [258]:
testdate = df.loc['2013-03-11':'2013-03-11']

In [259]:
testdate

Temperature  Wind Speed   Description
datetime                                                  
2013-03-11 00:00:00     289.6800         0.0  sky is clear
2013-03-11 01:00:00     289.7465         0.0  sky is clear
2013-03-11 02:00:00     287.2200         0.0  sky is clear
2013-03-11 03:00:00     286.1500         0.0  sky is clear
2013-03-11 04:00:00     284.9600         0.0  sky is clear
2013-03-11 05:00:00     284.4900         0.0  sky is clear
2013-03-11 06:00:00     283.8100         0.0  sky is clear
2013-03-11 07:00:00          NaN         0.0  sky is clear
2013-03-11 08:00:00          NaN         1.0  sky is clear
2013-03-11 09:00:00     282.5800         0.0  sky is clear
2013-03-11 10:00:00     281.8300         0.0  sky is clear
2013-03-11 11:00:00     281.6700         1.0  sky is clear
2013-03-11 12:00:00     281.3200         0.0  sky is clear
2013-03-11 13:00:00     281.0300         0.0  sky is clear
2013-03-11 14:00:00     280.7400         0.0  sky is clear
2013-03-11 15:00:00     282.2200         0.0  sky is clear
2013-03-11 16:00:00     283.1280         0.0  sky is clear
2013-03-11 17:00:00     288.6500         0.0  sky is clear
2013-03-11 18:00:00     290.1500         0.0  sky is clear
2013-03-11 19:00:00     291.6500         1.0  sky is clear
2013-03-11 20:00:00     293.1100         2.0  sky is clear
2013-03-11 21:00:00     293.8400         2.0  sky is clear
2013-03-11 22:00:00     293.4050         2.0  sky is clear
2013-03-11 23:00:00     292.9700         1.0  sky is clear

Find the average between 283.8100 and 282.5800. This is equal to 283.195. We'll then fill the null values with this average. Instead of deleting the rows we felt this was the best solution as we could need these rows later when comparing them to the NFL and NBA games. We felt getting the average between the row after and before the missing values would give us the best result. 

In [260]:
df['Temperature'] = df['Temperature'].fillna(283.195)

In [261]:
print(df[df["Temperature"].isnull()])

Empty DataFrame
Columns: [Temperature, Wind Speed, Description]
Index: []


We want to convert the temperature to degrees Celcius from kelvin. Celcius is more familiar to us and the general public so we felt it would make understanding easier. The formula to do this is T(°C)=T(K)-273.15, Therefore, we create a new column called 'Celcius' and use this formula. 

In [262]:
df["Celcius"] = df["Temperature"] - 273.15

In [263]:
df.head()

Temperature  Wind Speed   Description    Celcius
datetime                                                             
2012-10-01 13:00:00   291.870000         0.0          mist  18.720000
2012-10-01 14:00:00   291.868186         0.0  sky is clear  18.718186
2012-10-01 15:00:00   291.862844         0.0  sky is clear  18.712844
2012-10-01 16:00:00   291.857503         0.0  sky is clear  18.707503
2012-10-01 17:00:00   291.852162         0.0  sky is clear  18.702162

The Kelvin temperature column is now dropped as it is no longer needed

In [264]:
df = df.drop(df.columns[[0]], axis=1)

In [265]:
chiWea = df

In [266]:
chiWea.head()

Wind Speed   Description    Celcius
datetime                                                
2012-10-01 13:00:00         0.0          mist  18.720000
2012-10-01 14:00:00         0.0  sky is clear  18.718186
2012-10-01 15:00:00         0.0  sky is clear  18.712844
2012-10-01 16:00:00         0.0  sky is clear  18.707503
2012-10-01 17:00:00         0.0  sky is clear  18.702162

### The Weather Grading System

We now look at all the different types of Weather in terms of description. These will be grouped into different groups depending how good the weather is. 

In [267]:
pt = chiWea.groupby('Description')[('Celcius')].count()
pt.sort_values(ascending = False)

Description
sky is clear                    26136
haze                             3532
mist                             2961
broken clouds                    2565
overcast clouds                  2432
scattered clouds                 2271
light rain                       1949
few clouds                       1757
fog                               566
moderate rain                     481
smoke                             203
heavy intensity rain              127
light intensity drizzle           104
dust                               66
proximity thunderstorm             25
very heavy rain                    20
thunderstorm                       17
thunderstorm with light rain       13
shower rain                         8
drizzle                             5
proximity shower rain               4
light intensity shower rain         3
squalls                             3
thunderstorm with heavy rain        2
thunderstorm with rain              2
Name: Celcius, dtype: int64

We create the following function. If the description is 'sky is clear' or 'few clouds' this weather is considered good and is given a value of 3. Then if the description is 'overcast clouds', 'broken clouds' or 'scattered clouds' the weather is considered to be moderate and is given a value of 2. The rest (the bad weather) is given a value of 0. We wanted to get a clear difference between good and even moderate weather which is why bad weather is valued at 0 instead of 1. 

In [268]:
conditions = [
    (chiWea['Description'] == 'sky is clear') | (chiWea['Description'] == 'few clouds'),
    (chiWea['Description'] == 'overcast clouds') | (chiWea['Description'] == 'broken clouds') 
    | (chiWea['Description'] == 'scattered clouds') ]
choices = ['3', '2']
chiWea['DescRate'] = np.select(conditions, choices, default='0')
print(chiWea)

                     Wind Speed       Description    Celcius DescRate
datetime                                                             
2012-10-01 13:00:00         0.0              mist  18.720000        0
2012-10-01 14:00:00         0.0      sky is clear  18.718186        3
2012-10-01 15:00:00         0.0      sky is clear  18.712844        3
2012-10-01 16:00:00         0.0      sky is clear  18.707503        3
2012-10-01 17:00:00         0.0      sky is clear  18.702162        3
2012-10-01 18:00:00         0.0      sky is clear  18.696821        3
2012-10-01 19:00:00         0.0      sky is clear  18.691480        3
2012-10-01 20:00:00         0.0      sky is clear  18.686139        3
2012-10-01 21:00:00         0.0      sky is clear  18.680798        3
2012-10-01 22:00:00         0.0      sky is clear  18.675457        3
2012-10-01 23:00:00         0.0      sky is clear  18.670116        3
2012-10-02 00:00:00         0.0      sky is clear  18.664775        3
2012-10-02 01:00:00 

The next thing was to determine how good the temperatue was, so we could assign it a grade. We first got the mean temperature. Anything below the average temperature would be considered to be 'bad weather'. 

In [269]:
chiWea['Celcius'].mean()

17.695777983241285

In [287]:
chiWea['Celcius'].max()

42.32000000000005

In [288]:
chiWea['Celcius'].min()

-6.6463333329999728

Again the same type of code is used as above. This time if the temperature is below or equal to 10 a value of 0 is assigned as it is considered bad weather. Anything above or equal to 20 degrees celcius is given a value of 3 as it is good weather. The In between temperatures are considered moderate and is given a value of 2. 

In [270]:
conditions = [
    (chiWea['Celcius'] <= 13),
    (chiWea['Celcius'] >= 23)]
choices = ['0', '3']
chiWea['TempRate'] = np.select(conditions, choices, default='2')
print(chiWea)

                     Wind Speed       Description    Celcius DescRate TempRate
datetime                                                                      
2012-10-01 13:00:00         0.0              mist  18.720000        0        2
2012-10-01 14:00:00         0.0      sky is clear  18.718186        3        2
2012-10-01 15:00:00         0.0      sky is clear  18.712844        3        2
2012-10-01 16:00:00         0.0      sky is clear  18.707503        3        2
2012-10-01 17:00:00         0.0      sky is clear  18.702162        3        2
2012-10-01 18:00:00         0.0      sky is clear  18.696821        3        2
2012-10-01 19:00:00         0.0      sky is clear  18.691480        3        2
2012-10-01 20:00:00         0.0      sky is clear  18.686139        3        2
2012-10-01 21:00:00         0.0      sky is clear  18.680798        3        2
2012-10-01 22:00:00         0.0      sky is clear  18.675457        3        2
2012-10-01 23:00:00         0.0      sky is clear  1

In [271]:
chiWea.head()

Wind Speed   Description    Celcius DescRate TempRate
datetime                                                                  
2012-10-01 13:00:00         0.0          mist  18.720000        0        2
2012-10-01 14:00:00         0.0  sky is clear  18.718186        3        2
2012-10-01 15:00:00         0.0  sky is clear  18.712844        3        2
2012-10-01 16:00:00         0.0  sky is clear  18.707503        3        2
2012-10-01 17:00:00         0.0  sky is clear  18.702162        3        2

The average wind speed is found. This is relatively low. Although there is no indication to what the wind speed is measured in we are inclined to think the measurement is knots. The average wind speed in Los Angeles is around 6-10km per hour according to https://www.isws.illinois.edu/statecli/wind/wind.htm . This average in knots would fit into this bracket. 

In [272]:
chiWea['Wind Speed'].mean()

1.2195483072571378

Wind speed less than or equal to 2 is considered to be very calm and is considered good. It is valued at 3. Between 2 and 5 is considered moderate and is valued at 2. Above 5 knots is considered bad and is given a value of 0. 

In [273]:
conditions = [
    (chiWea['Wind Speed'] == 0),
    (chiWea['Wind Speed'] > 0) | (chiWea['Wind Speed'] < 5) ]
choices = ['3', '2']
chiWea['WindRate'] = np.select(conditions, choices, default='0')
print(chiWea)

                     Wind Speed       Description    Celcius DescRate  \
datetime                                                                
2012-10-01 13:00:00         0.0              mist  18.720000        0   
2012-10-01 14:00:00         0.0      sky is clear  18.718186        3   
2012-10-01 15:00:00         0.0      sky is clear  18.712844        3   
2012-10-01 16:00:00         0.0      sky is clear  18.707503        3   
2012-10-01 17:00:00         0.0      sky is clear  18.702162        3   
2012-10-01 18:00:00         0.0      sky is clear  18.696821        3   
2012-10-01 19:00:00         0.0      sky is clear  18.691480        3   
2012-10-01 20:00:00         0.0      sky is clear  18.686139        3   
2012-10-01 21:00:00         0.0      sky is clear  18.680798        3   
2012-10-01 22:00:00         0.0      sky is clear  18.675457        3   
2012-10-01 23:00:00         0.0      sky is clear  18.670116        3   
2012-10-02 00:00:00         0.0      sky is clear  

In [274]:
chiWea.head()

Wind Speed   Description    Celcius DescRate TempRate  \
datetime                                                                     
2012-10-01 13:00:00         0.0          mist  18.720000        0        2   
2012-10-01 14:00:00         0.0  sky is clear  18.718186        3        2   
2012-10-01 15:00:00         0.0  sky is clear  18.712844        3        2   
2012-10-01 16:00:00         0.0  sky is clear  18.707503        3        2   
2012-10-01 17:00:00         0.0  sky is clear  18.702162        3        2   

                    WindRate  
datetime                      
2012-10-01 13:00:00        3  
2012-10-01 14:00:00        3  
2012-10-01 15:00:00        3  
2012-10-01 16:00:00        3  
2012-10-01 17:00:00        3

In [275]:
chiWea.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 45252 entries, 2012-10-01 13:00:00 to 2017-11-30 00:00:00
Freq: H
Data columns (total 6 columns):
Wind Speed     45252 non-null float64
Description    45252 non-null object
Celcius        45252 non-null float64
DescRate       45252 non-null object
TempRate       45252 non-null object
WindRate       45252 non-null object
dtypes: float64(2), object(4)
memory usage: 3.7+ MB


We have to change our new columns to int as we are going to add them together to get our overall weather rating. 

In [276]:
chiWea['DescRate'] = chiWea['DescRate'].astype(int)

In [277]:
chiWea['TempRate'] = chiWea['TempRate'].astype(int)

In [278]:
chiWea['WindRate'] = chiWea['WindRate'].astype(int)

In [279]:
chiWea.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 45252 entries, 2012-10-01 13:00:00 to 2017-11-30 00:00:00
Freq: H
Data columns (total 6 columns):
Wind Speed     45252 non-null float64
Description    45252 non-null object
Celcius        45252 non-null float64
DescRate       45252 non-null int32
TempRate       45252 non-null int32
WindRate       45252 non-null int32
dtypes: float64(2), int32(3), object(1)
memory usage: 3.1+ MB


The Overall column is created it consits of the Description Rating, Temperature Rating and Wind Rating added together. This overall rating gives us a sense of how good the weather was for that hour. 

In [280]:
chiWea['Overall'] = chiWea["DescRate"] + chiWea["TempRate"] + chiWea["WindRate"]

In [281]:
chiWea.head()

Wind Speed   Description    Celcius  DescRate  TempRate  \
datetime                                                                       
2012-10-01 13:00:00         0.0          mist  18.720000         0         2   
2012-10-01 14:00:00         0.0  sky is clear  18.718186         3         2   
2012-10-01 15:00:00         0.0  sky is clear  18.712844         3         2   
2012-10-01 16:00:00         0.0  sky is clear  18.707503         3         2   
2012-10-01 17:00:00         0.0  sky is clear  18.702162         3         2   

                     WindRate  Overall  
datetime                                
2012-10-01 13:00:00         3        5  
2012-10-01 14:00:00         3        8  
2012-10-01 15:00:00         3        8  
2012-10-01 16:00:00         3        8  
2012-10-01 17:00:00         3        8

We then want to group the weather into Good, Moderate and Overall Weather. 8 and over is Good. 6 to 8 is considered moderate. And under 6 would be considered to be Bad. For example if we have a case where the temperature is 21 degrees, overcast clouds and rain, this would be considered bad as Raining is always considered bad in terms of weather, and prevents people from going outside. 

In [282]:
conditions = [
    (chiWea['Overall'] >= 8), 
    (chiWea['Overall'] >= 6) & (chiWea['Overall'] < 8) ]
choices = ['Good', 'Moderate']
chiWea['Weather'] = np.select(conditions, choices, default='Bad')

In [283]:
chiWea.head(50)

Wind Speed   Description    Celcius  DescRate  TempRate  \
datetime                                                                       
2012-10-01 13:00:00         0.0          mist  18.720000         0         2   
2012-10-01 14:00:00         0.0  sky is clear  18.718186         3         2   
2012-10-01 15:00:00         0.0  sky is clear  18.712844         3         2   
2012-10-01 16:00:00         0.0  sky is clear  18.707503         3         2   
2012-10-01 17:00:00         0.0  sky is clear  18.702162         3         2   
2012-10-01 18:00:00         0.0  sky is clear  18.696821         3         2   
2012-10-01 19:00:00         0.0  sky is clear  18.691480         3         2   
2012-10-01 20:00:00         0.0  sky is clear  18.686139         3         2   
2012-10-01 21:00:00         0.0  sky is clear  18.680798         3         2   
2012-10-01 22:00:00         0.0  sky is clear  18.675457         3         2   
2012-10-01 23:00:00         0.0  sky is clear  18.670116         3         2   
2012-10-02 00:00:00         0.0  sky is clear  18.664775         3         2   
2012-10-02 01:00:00         0.0  sky is clear  18.659434         3         2   
2012-10-02 02:00:00         0.0  sky is clear  18.654093         3         2   
2012-10-02 03:00:00         0.0  sky is clear  18.648752         3         2   
2012-10-02 04:00:00         0.0  sky is clear  18.643411         3         2   
2012-10-02 05:00:00         0.0  sky is clear  18.638070         3         2   
2012-10-02 06:00:00         0.0  sky is clear  18.632729         3         2   
2012-10-02 07:00:00         0.0  sky is clear  18.627387         3         2   
2012-10-02 08:00:00         0.0  sky is clear  18.622046         3         2   
2012-10-02 09:00:00         0.0  sky is clear  18.616705         3         2   
2012-10-02 10:00:00         0.0  sky is clear  18.611364         3         2   
2012-10-02 11:00:00         0.0  sky is clear  18.606023         3         2   
2012-10-02 12:00:00         0.0  sky is clear  18.600682         3         2   
2012-10-02 13:00:00         0.0  sky is clear  18.595341         3         2   
2012-10-02 14:00:00         0.0          mist  18.590000         0         2   
2012-10-02 15:00:00         0.0           fog  20.820000         0         2   
2012-10-02 16:00:00         0.0  sky is clear  22.440000         3         2   
2012-10-02 17:00:00         1.0  sky is clear  24.780000         3         3   
2012-10-02 18:00:00         0.0  sky is clear  30.680000         3         3   
2012-10-02 19:00:00         1.0  sky is clear  31.880000         3         3   
2012-10-02 20:00:00         1.0  sky is clear  32.970000         3         3   
2012-10-02 21:00:00         2.0  sky is clear  34.510000         3         3   
2012-10-02 22:00:00         3.0  sky is clear  34.560000         3         3   
2012-10-02 23:00:00         1.0  sky is clear  33.720000         3         3   
2012-10-03 00:00:00         2.0  sky is clear  32.550000         3         3   
2012-10-03 01:00:00         2.0  sky is clear  30.540000         3         3   
2012-10-03 02:00:00         1.0  sky is clear  28.870000         3         3   
2012-10-03 03:00:00         0.0  sky is clear  27.270000         3         3   
2012-10-03 04:00:00         0.0  sky is clear  26.490000         3         3   
2012-10-03 05:00:00         0.0  sky is clear  26.470000         3         3   
2012-10-03 06:00:00         0.0  sky is clear  23.230000         3         3   
2012-10-03 07:00:00         0.0  sky is clear  22.050000         3         2   
2012-10-03 08:00:00         0.0  sky is clear  22.050000         3         2   
2012-10-03 09:00:00         0.0  sky is clear  21.510000         3         2   
2012-10-03 10:00:00         0.0  sky is clear  21.070000         3         2   
2012-10-03 11:00:00         0.0  sky is clear  20.710000         3         2   
2012-10-03 12:00:00         0.0  sky is clear  20.085000         3         2   
2012-10-03 13:00:00  

In [284]:
chiWea.sort_values(by=['Overall'])

Wind Speed      Description    Celcius  DescRate  \
datetime                                                                
2016-01-24 12:00:00         1.0             mist   9.710000         0   
2017-02-28 05:00:00         1.0             mist  11.420000         0   
2017-02-28 06:00:00         1.0             mist  10.680000         0   
2017-01-20 06:00:00         1.0             mist  12.380000         0   
2016-01-05 15:00:00         3.0             mist  12.210000         0   
2017-01-20 05:00:00         4.0             mist  12.580000         0   
2016-01-05 16:00:00         2.0    moderate rain  12.194844         0   
2016-01-05 17:00:00         2.0              fog  12.140000         0   
2016-01-05 18:00:00         1.0             mist  12.560000         0   
2016-01-05 19:00:00         4.0              fog  12.730000         0   
2017-02-28 07:00:00         3.0             mist  10.740000         0   
2016-01-05 22:00:00         1.0              fog  12.840000         0   
2016-01-05 23:00:00         1.0  very heavy rain  10.591271         0   
2017-01-20 07:00:00         1.0             mist  12.240000         0   
2016-01-06 00:00:00         3.0    moderate rain  12.670000         0   
2017-02-28 09:00:00         1.0             mist   9.820000         0   
2017-02-28 10:00:00         1.0             mist   9.400000         0   
2017-02-28 11:00:00         1.0             mist   8.900000         0   
2017-02-28 12:00:00         1.0             mist   8.490000         0   
2017-02-28 13:00:00         1.0             mist   8.070000         0   
2017-02-28 14:00:00         1.0             mist   7.570000         0   
2017-02-28 15:00:00         1.0             mist   7.350000         0   
2017-02-28 16:00:00         2.0             mist   9.990000         0   
2017-02-28 17:00:00         2.0             haze  12.060000         0   
2013-01-27 04:00:00         1.0       light rain  12.810000         0   
2013-01-27 05:00:00         1.0       light rain  12.618000         0   
2016-03-22 14:00:00        15.0             haze  12.020000         0   
2017-02-28 08:00:00         1.0             mist  10.490000         0   
2015-12-09 16:00:00         1.0              fog  11.920000         0   
2017-02-28 04:00:00         1.0              fog  12.070000         0   
...                         ...              ...        ...       ...   
2013-07-01 02:00:00         0.0     sky is clear  29.930000         3   
2013-07-01 03:00:00         0.0     sky is clear  28.980000         3   
2013-08-31 16:00:00         0.0     sky is clear  25.851333         3   
2013-08-31 15:00:00         0.0       few clouds  25.770500         3   
2016-09-27 00:00:00         0.0     sky is clear  34.700000         3   
2017-10-06 03:00:00         0.0     sky is clear  23.470000         3   
2013-06-29 04:00:00         0.0     sky is clear  23.710000         3   
2013-06-29 03:00:00         0.0     sky is clear  25.600000         3   
2016-09-26 02:00:00         0.0     sky is clear  31.660000         3   
2016-09-26 03:00:00         0.0     sky is clear  29.510000         3   
2016-09-26 05:00:00         0.0     sky is clear  26.340000         3   
2013-09-03 06:00:00         0.0     sky is clear  23.580000         3   
2013-09-03 05:00:00         0.0     sky is clear  23.790000         3   
2013-09-03 03:00:00         0.0     sky is clear  25.880000         3   
2016-09-26 09:00:00         0.0     sky is clear  23.330000         3   
2016-09-26 10:00:00         0.0     sky is clear  23.333272         3   
2017-10-07 04:00:00         0.0     sky is clear  24.810000         3   
2013-06-28 03:00:00         0.0     sky is clear  25.460000         3   
2017-10-07 03:00:00         0.0     sky is clear  27.060000         3   
2017-10-06 21:00:00         0.0     sky is clear  31.540000         3   
2014-08-16 04:00:00         0.0     sky is clear  23.112000         3   
2016-07-18 04:00:00         0.0     sky is clear  24.800000         3   
2016-0

We give an overall count of the weather grades. Unsurprsingly, there is more bad weather than either moderate or good. This would be because a lot of the hours would be overnight when the temperature is lower and the sun is in. 

In [285]:
wea1 = chiWea.groupby('Weather')[('Overall')].count()
wea1.sort_values(ascending=False)

Weather
Moderate    16769
Bad         15167
Good        13316
Name: Overall, dtype: int64

We save this to a new csv, where it will be used for analysis. 

In [286]:
chiWea.to_csv('../../data/prep/500_LAWea.csv')